src/Oscoin/Data/OscoinTx.hs

```diff
@@ -79,6 +115,61 @@ data TxMessage c =
     -- ^ Transfer balance between two accounts.
     deriving (Generic)
 
+instance
+    ( Serialise (AccountId c)
+    , Serialise (Crypto.Hash c)
+    , Serialise (Contribution c)
+    , Serialise (DependencyUpdate c)
+    ) => Serialise (TxMessage c)
+  where
+    encode (TxRegisterProject acc) =
+           CBOR.encodeListLen 2
+        <> CBOR.encodeWord    0
+        <> CBOR.encode        acc
+    encode (TxUnregisterProject acc) =
+           CBOR.encodeListLen 2
+        <> CBOR.encodeWord    1
+        <> CBOR.encode        acc
+    encode (TxAuthorize proj acc) =
+           CBOR.encodeListLen 3
+        <> CBOR.encodeWord    2
+        <> CBOR.encode        proj
+        <> CBOR.encode        acc
+    encode (TxDeauthorize proj acc) =
+           CBOR.encodeListLen 3
+        <> CBOR.encodeWord    3
+        <> CBOR.encode        proj
+        <> CBOR.encode        acc
+    encode (TxCheckpoint proj hsh cs ds) =
+           CBOR.encodeListLen 5
+        <> CBOR.encodeWord    4
+        <> CBOR.encode        proj
+        <> CBOR.encode        hsh
+        <> CBOR.encode        cs
+        <> CBOR.encode        ds
+    encode (TxUpdateContract proj) =
+           CBOR.encodeListLen 2
+        <> CBOR.encodeWord    5
+        <> CBOR.encode        proj
+    encode (TxTransfer acc bal) =
+           CBOR.encodeListLen 3
+        <> CBOR.encodeWord    6
+        <> CBOR.encode        acc
+        <> CBOR.encode        bal
+
+    decode = do
+        pre <- liftA2 (,) CBOR.decodeListLen CBOR.decodeWord
+        case pre of
+            (2, 0) -> TxRegisterProject   <$> CBOR.decode
+            (2, 1) -> TxUnregisterProject <$> CBOR.decode
+            (3, 2) -> TxAuthorize         <$> CBOR.decode <*> CBOR.decode
+            (3, 3) -> TxDeauthorize       <$> CBOR.decode <*> CBOR.decode
+            (5, 4) -> TxCheckpoint        <$> CBOR.decode <*> CBOR.decode <*> CBOR.decode <*> CBOR.decode
+            (2, 5) -> TxUpdateContract    <$> CBOR.decode
+            (3, 6) -> TxTransfer          <$> CBOR.decode <*> CBOR.decode
```

Comment:
Same as above -- perhaps it would help with more precise error reporting if you tuck the list-len checking inside each branch, after decoding the tag.